server/src/main/java/org/elasticsearch/cluster/node/VersionInformation.java

```diff
@@ -10,25 +10,54 @@
 package org.elasticsearch.cluster.node;
 
 import org.elasticsearch.Version;
+import org.elasticsearch.env.BuildVersion;
 import org.elasticsearch.index.IndexVersion;
 import org.elasticsearch.index.IndexVersions;
 
 import java.util.Objects;
 
 /**
  * Represents the versions of various aspects of an Elasticsearch node.
- * @param nodeVersion       The node {@link Version}
+ * @param buildVersion      The node {@link BuildVersion}
  * @param minIndexVersion   The minimum {@link IndexVersion} supported by this node
  * @param maxIndexVersion   The maximum {@link IndexVersion} supported by this node
  */
-public record VersionInformation(Version nodeVersion, IndexVersion minIndexVersion, IndexVersion maxIndexVersion) {
+public record VersionInformation(
+    BuildVersion buildVersion,
+    Version nodeVersion,
+    IndexVersion minIndexVersion,
+    IndexVersion maxIndexVersion
+) {
 
     public static final VersionInformation CURRENT = new VersionInformation(
-        Version.CURRENT,
+        BuildVersion.current(),
         IndexVersions.MINIMUM_COMPATIBLE,
         IndexVersion.current()
     );
 
+    public VersionInformation {
+        Objects.requireNonNull(buildVersion);
+        Objects.requireNonNull(minIndexVersion);
+        Objects.requireNonNull(maxIndexVersion);
+    }
+
+    public VersionInformation(BuildVersion version, IndexVersion minIndexVersion, IndexVersion maxIndexVersion) {
+        this(version, Version.CURRENT, minIndexVersion, maxIndexVersion);
+        /*
+         * Whilst DiscoveryNode.getVersion exists, we need to be able to get a Version from VersionInfo
+         * This needs to be consistent - on serverless, BuildVersion has an id of -1, which translates
+         * to a nonsensical Version. So all consumers of Version need to be moved to BuildVersion
+         * before we can remove Version from here.
+         */
+        // for the moment, check this is only called with current() so the implied Version is correct
+        assert version.equals(BuildVersion.current()) : version + " is not " + BuildVersion.current();
```

Comment:
Won't this fail during an upgrade? The deprecated ctor below calls fromVersionId, which may not be current

Reply:
This check is only for the overload with `BuildVersion` only. My plan is that once nothing is referencing `VersionInfo.nodeVersion` we can remove the `Version` field completely. This particular constructor is here now for specific instances which we can just change over to `BuildVersion.current()` straightaway, easily removing more references to `Version`. Any more complicated uses will need to use the `Version` or `BuildVersion, Version` overload for the meantime.